inst/include/utils.h

```diff
@@ -1,5 +1,75 @@
 #include <torch.h>
 
+#include <stack>
+#include <mutex>
+#include <condition_variable>
+#include <future>
+#include <functional>
+#include <utility>
+#include <type_traits>
+
+template <typename F>
+class ScopeGuard {
```

Comment:
I expect this to be part of the C++ standard in near future (e.g., see https://en.cppreference.com/w/cpp/experimental/scope_exit).

This is just my straw-man implementation of the scope guard concept.
There is also a much more advanced version implemented by advanced C++ gurus doing advanced C++ full-time in https://github.com/facebook/folly/blob/47eda458e9763093e0bc1baa3bebef6aeb816513/folly/ScopeGuard.h#L1